src/gui/SampleThumbnail.cpp

```diff
@@ -0,0 +1,246 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program (see COPYING); if not, write to the
+ * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301 USA.
+ *
+ */
+
+#include "SampleThumbnail.h"
+#include "Sample.h"
+
+namespace lmms {
+
+SampleThumbnail::Bit::Bit(const SampleFrame& frame)
+	: max(std::max(frame.left(), frame.right()))
+	, min(std::min(frame.left(), frame.right()))
+{}
+
+SampleThumbnail::Bit SampleThumbnail::Bit::merge(const Bit& a, const Bit& b)
+{
+	Bit out{};
+	out.min = std::min(a.min, b.min);
+	out.max = std::max(a.max, b.max);
+	return out;
+}
+
+SampleThumbnail::Bit SampleThumbnail::Bit::merge(const Bit& current, const SampleFrame& frame)
+{
+	return merge(current, Bit{frame});
+}
+
+static SampleThumbnail::Bit mergeBits(
+	SampleThumbnail::Thumbnail::const_iterator start,
+	SampleThumbnail::Thumbnail::const_iterator end,
+	SampleThumbnail::Bit init_value = SampleThumbnail::Bit{}
+) {
+	return std::accumulate(
+		start,
+		end,
+		init_value,
+		[&](const SampleThumbnail::Bit& a, const SampleThumbnail::Bit& b) { return SampleThumbnail::Bit::merge(a, b); }
+	);
+}
+
+static SampleThumbnail::Bit mergeFrames(
+	SampleBuffer::const_iterator start,
+	SampleBuffer::const_iterator end
+) {
+	return std::accumulate(
+		start,
+		end,
+		SampleThumbnail::Bit{},
+		[&](const SampleThumbnail::Bit& a, const SampleFrame& b) { return SampleThumbnail::Bit::merge(a, b); }
+	);
+}
+
+/* DEPRECATED; functionality is kept for testing conveniences */
+bool SampleThumbnail::selectFromGlobalThumbnailMap(const Sample& inputSample)
+{
+	const auto samplePtr = inputSample.buffer();
+	const auto name = inputSample.sampleFile();
+	const auto end = s_sampleThumbnailCacheMap.end();
+
+	if (const auto list = s_sampleThumbnailCacheMap.find(name); list != end)
+	{
+		m_thumbnailCache = list->second;
+		return true;
+	}
+
+	m_thumbnailCache = std::make_shared<ThumbnailCache>();
+	s_sampleThumbnailCacheMap.insert(std::make_pair(name, m_thumbnailCache));
+	return false;
+}
+
+/* DEPRECATED; functionality is kept for testing conveniences */
+void SampleThumbnail::cleanUpGlobalThumbnailMap()
+{
+	auto map = s_sampleThumbnailCacheMap.begin();
+	while (map != s_sampleThumbnailCacheMap.end())
+	{
+		// All sample thumbnails are destroyed, a.k.a sample goes out of use
+		if (map->second.use_count() == 1)
+		{
+			s_sampleThumbnailCacheMap.erase(map);
+			map = s_sampleThumbnailCacheMap.begin();
+			continue;
+		}
+
+		map++;
+	}
+}
+
+SampleThumbnail::Thumbnail SampleThumbnail::generate(const std::size_t thumbnailSize, const Sample& sample, const std::size_t size)
```

Comment:
Passing in `sample` is unnecessary, as you only access the buffer inside each `sample`. You don't need to carry all of that data around (which can hurt performance as it can cause a higher chance for cache misses). Just take in a `SampleBuffer*` or `SampleFrame*`.